nexus/sdk-python.md

```diff
@@ -0,0 +1,330 @@
+# Python Nexus design proposal
+
+This is an early-stage design document for how Nexus functionality will be exposed in the Nexus Python SDK and in the Temporal Python SDK. The design
+is still evolving, and feedback is welcome.
+
+The design proposed here is being prototyped in this draft PR:
+[https://github.com/temporalio/samples-python/pull/174](https://github.com/temporalio/samples-python/pull/174)
+
+# Introduction
+
+A Nexus service is a collection of Nexus operations, each of which supports the four verbs `start`, `fetch_info`, `fetch_result`, and `cancel`.
+
+On the handler side of Nexus, users need to be able to:
+
+1. **Define a service interface**
+    
+    This will be consumed by caller code to create a type-safe client, and by the service implementation to verify that the interface has been implemented.
+    
+2. **Implement an operation in a service**
+
+This document focuses on these two areas (they’re the most complex aspects of Nexus SDK design).
+
+Note that this document discusses Temporal-specific Nexus at the end. Everything before that is intended to stand on its own, i.e. the part before the Temporal section is intended to make sense without any implication that Temporal is in the background.
+
+# 1. Defining a service interface
+
+The team writing the Nexus service will define the interface in Python (This Python definition will be the target of codegen from future Nexus service IDL).
+
+The interface will be defined in a dataclass-like style [thanks @cretz for this idea]:
+
+```python
+import nexusrpc
+
+@nexusrpc.service
+class MyNexusService:
+    echo: nexusrpc.Operation[EchoInput, EchoOutput]
+    hello: nexusrpc.Operation[HelloInput, HelloOutput]
+```
+
+Note that:
+
+- What the user actually writes here defines the operations as purely type-level constructs: the decorator creates `NexusOperation` instances at import time that can be referenced by the caller-side Nexus service client.
+
+The interface will be used by the caller as follows:
+
+```python
+from my_nexus_service import EchoInput, EchoOutput, MyNexusService
+
+@workflow.defn
+class EchoCallerWorkflow:
+    def __init__(self):
+        self.nexus_client = NexusClient(
+            MyNexusService,  # or string name "my-nexus-service",
+            "my-nexus-endpoint-name",
+            schedule_to_close_timeout=timedelta(seconds=30),
```

Comment:
I think this needs to be per-operation kwarg. Java uniquely does some of these at the service/stub level because of how it proxies, but all other SDKs can do operation level.